### PR TITLE
[carla] In carla example, save all images and measurements to local disk

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Saves all images (in 400x300) to local disk, along with measurements, action, controls, and reward. The directory structure is as follows:

```
CARLA_OUT/
   CameraRGB/
      <episode_id>_<step>.jpg
      ...
   CameraDepth/
      <episode_id>_<step>.jpg
      ...
    measurements_<episode_id>.json
    ...
```

### CODE REVIEW

File: examples/carla/env.py [hunk 1]
```diff
@@ -206,38 +231,77 @@ def _step(self, action):
...
+                interpolation=cv2.INTER_AREA)
         else:
-            return image.data.reshape(
-                self.config["x_res"], self.config["y_res"], 3)
+            data = image.data.reshape(
+                self.config["render_y_res"], self.config["render_x_res"], 3)
+            data = cv2.resize(
+                data, (self.config["x_res"], self.config["y_res"]),
+                interpolation=cv2.INTER_AREA)
+            data = (data.astype(np.float32) - 128) / 128
```
Comment (ericl) [inline examples/carla/env.py]: Tested both out by using scipy.imsave on the results, and they appear correct.

File: examples/carla/env.py [hunk 2]
```diff
@@ -206,38 +231,77 @@ def _step(self, action):
...
+            data = data.reshape(
+                self.config["render_y_res"], self.config["render_x_res"], 1)
+            data = cv2.resize(
+                data, (self.config["x_res"], self.config["y_res"]),
+                interpolation=cv2.INTER_AREA)
         else:
-            return image.data.reshape(
-                self.config["x_res"], self.config["y_res"], 3)
+            data = image.data.reshape(
+                self.config["render_y_res"], self.config["render_x_res"], 3)
```
Comment (ericl) [inline examples/carla/env.py]: Note: (y, x)

**Comment by danielsuo** [comment]
Awesome work! Quick question: presumably, we're interested in saving the data to local disk for later batch processing? If so, we may care less about real-time performance and more about resolution.

In the [private] BDD repo, we were able to get 1920x1080 by dumping to a sequential binary file (i.e., no processing) at ~12 fps on Titan X Maxwell (close to the [max](https://github.com/carla-simulator/carla/issues/10) others reported). Of course, we need some post-processing to convert that into whatever usable format after the fact.

**Comment by ericl** [comment]
Yeah, the motivation here is two fold: off policy RL algorithms can be bootstrapped off of existing data, and also dumping makes it easier to debug training progress.

I chose 400x300 to save on disk space -- in your binary format, what's the space usage like? If the code can be shared it might make sense to also save in that format as well for more flexibility in the future.

**Comment by danielsuo** [comment]
I see, makes sense. I've been advised by @fyu that it'd be better to have higher-resolution data, but I fully admit I don't know all of the factors to consider.

Our binary format is really straightforward---for each of the width x height x channels elements in each raw data array, allocate 4 bytes (whether integer or floating point). In the future, we can think about quantizing / using variable-length integers or something like that, but keeping simple for now.

**Comment by danielsuo** [comment]
I see, makes sense. I've been advised by @fyu that higher resolution is important, but I'll fully admit that I don't know all the factors to consider.

I don't know that we're allowed to share the code publicly, but the format for images is really simple: one big file with 4 bytes for each element in the width x height x channels x sensors x time array. We have some code to process this into human-viewable images (e.g., RGB image for scene; log-scale grayscale depth), but reading in the single binary file during replay /training is also much faster than opening a ton of small files.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**